users/postgres/groups.go

```diff
@@ -0,0 +1,445 @@
+		return errors.Wrap(errUpdateDB, err)
+	}
+
+	return nil
+}
+
+func (gr groupRepository) Delete(ctx context.Context, groupID string) error {
+	qd := `DELETE FROM groups WHERE id = :id`
+	dbg, err := toDBGroup(users.Group{ID: groupID})
+	if err != nil {
+		return errors.Wrap(errUpdateDB, err)
+	}
+	res, err := gr.db.NamedExecContext(ctx, qd, dbg)
+	if err != nil {
+		return errors.Wrap(errDeleteGroupDB, err)
+	}
+
+	cnt, err := res.RowsAffected()
+	if err != nil {
+		return errors.Wrap(errDeleteGroupDB, err)
+	}
+
+	if cnt != 1 {
+		return errors.Wrap(users.ErrDeleteGroupMissing, err)
+	}
+	return nil
+}
+
+func (gr groupRepository) RetrieveByID(ctx context.Context, id string) (users.Group, error) {
+	q := `SELECT id, name, owner_id, parent_id, description, metadata FROM groups WHERE id = $1`
+	dbu := dbGroup{
+		ID: id,
+	}
+	if err := gr.db.QueryRowxContext(ctx, q, id).StructScan(&dbu); err != nil {
+		if err == sql.ErrNoRows {
+			return users.Group{}, errors.Wrap(users.ErrNotFound, err)
+
+		}
+		return users.Group{}, errors.Wrap(errRetrieveDB, err)
+	}
+	return toGroup(dbu), nil
+}
+
+func (gr groupRepository) RetrieveByName(ctx context.Context, name string) (users.Group, error) {
+	q := `SELECT id, name, description, metadata FROM groups WHERE name = $1`
+
+	dbu := dbGroup{
+		Name: name,
+	}
+	if err := gr.db.QueryRowxContext(ctx, q, name).StructScan(&dbu); err != nil {
+		if err == sql.ErrNoRows {
+			return users.Group{}, errors.Wrap(users.ErrNotFound, err)
+
+		}
+		return users.Group{}, errors.Wrap(errRetrieveDB, err)
+	}
+	group := toGroup(dbu)
+	return group, nil
+}
+
+func (gr groupRepository) RetrieveAll(ctx context.Context, groupID string, offset, limit uint64, gm users.Metadata) (users.GroupPage, error) {
+	_, mq, err := getGroupsMetadataQuery(gm)
+	if err != nil {
+		return users.GroupPage{}, errors.Wrap(errRetrieveDB, err)
+	}
+	q := fmt.Sprintf(`WITH RECURSIVE subordinates AS (
+						SELECT id, owner_id, parent_id, name, description, metadata
+						FROM groups
+						WHERE id = :id 
+						UNION
+							SELECT groups.id, groups.owner_id, groups.parent_id, groups.name, groups.description, groups.metadata
+							FROM groups 
+							INNER JOIN subordinates s ON s.id = groups.parent_id %s
+					) SELECT * FROM subordinates ORDER BY id LIMIT :limit OFFSET :offset`, mq)
+
+	dbPage, err := toDBGroupPage("", groupID, offset, limit, gm)
+	if err != nil {
+		return users.GroupPage{}, errors.Wrap(errSelectDb, err)
+	}
+	rows, err := gr.db.NamedQueryContext(ctx, q, dbPage)
+	if err != nil {
+		return users.GroupPage{}, errors.Wrap(errSelectDb, err)
+	}
+	defer rows.Close()
+
+	var items []users.Group
+	for rows.Next() {
+		dbgr := dbGroup{}
+		if err := rows.StructScan(&dbgr); err != nil {
+			return users.GroupPage{}, errors.Wrap(errSelectDb, err)
+		}
+
+		gr := toGroup(dbgr)
+		if err != nil {
+			return users.GroupPage{}, err
+		}
+
+		items = append(items, gr)
+	}
+
+	cq := fmt.Sprintf(`WITH RECURSIVE subordinates AS (
```

Comment:
I'm not sure if we need this. I prefer just adding a limit to the depth to some arbitrary number (10?) and return the list with no need for pagination and complications. What do you think? I doubt that we'll ever need more than 10-15 ancestors, and we don't need pagination for that.

Reply:
This implies that we'll need this COUNT to check the depth when adding a group.

Reply:
it works like a charm

Reply:
Do you want to keep it like this? I'm perfectly fine with it. @mainflux/maintainers Your opinion? It's `ancestors pagination` VS `ancestors depth limit`.